From 5c60c28819655e6193941f0d6847794bb7fa24e2 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Thu, 30 Sep 2021 13:51:58 +0200
Subject: [PATCH 0222/1695] drivers: stm32_firewall: add a firewall interface

Add a STM32 global interface to register firewall controller.
It defines generic operations to manage platform security
configuration.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 944af8cef26a04c43a1177b95645a90ac9318639 drivers: stm32_firewall: add a firewall interface]
Change-Id: Ie7a5e88baf20e189fc3fbe1516443026872f3097
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
---
 core/drivers/firewall/stm32_firewall.c | 191 ++++++++++++++++++++++
 core/drivers/firewall/sub.mk           |   1 +
 core/drivers/sub.mk                    |   1 +
 core/include/drivers/stm32_firewall.h  | 212 +++++++++++++++++++++++++
 4 files changed, 405 insertions(+)
 create mode 100644 core/drivers/firewall/stm32_firewall.c
 create mode 100644 core/drivers/firewall/sub.mk
 create mode 100644 core/include/drivers/stm32_firewall.h

diff --git a/core/drivers/firewall/stm32_firewall.c b/core/drivers/firewall/stm32_firewall.c
new file mode 100644
index 000000000..6ad09937c
--- /dev/null
+++ b/core/drivers/firewall/stm32_firewall.c
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2021, STMicroelectronics - All Rights Reserved
+ */
+
+#include <assert.h>
+#include <drivers/stm32_firewall.h>
+#include <kernel/boot.h>
+#include <kernel/dt_driver.h>
+#include <kernel/spinlock.h>
+#include <libfdt.h>
+#include <util.h>
+
+static LIST_HEAD(fw_dev_list_head, stm32_firewall_device) fw_dev_list =
+	LIST_HEAD_INITIALIZER(fw_dev_list_head);
+
+static unsigned int list_lock = SPINLOCK_UNLOCK;
+
+static struct stm32_firewall_device *find_device(paddr_t base, size_t size,
+						 unsigned int *idx)
+{
+	struct stm32_firewall_device *fdev = NULL;
+	unsigned int i = 0;
+
+	assert(idx);
+
+	LIST_FOREACH(fdev, &fw_dev_list, dev_list) {
+		for (i = 0; i < fdev->compat->compat_size; i++) {
+			struct stm32_firewall_reg *reg = fdev->compat->reg + i;
+
+			if (!reg->addr && !reg->size)
+				continue;
+
+			if (!reg->size) {
+				if (base != reg->addr)
+					continue;
+
+				if (size)
+					DMSG("IOMEM %#"PRIxPA": non-null size",
+					     base);
+
+				*idx = i;
+				return fdev;
+			}
+
+			if (reg->addr && reg->size && base >= reg->addr &&
+			    base + size - 1 <= reg->addr + reg->size - 1) {
+				*idx = i;
+				return fdev;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+TEE_Result stm32_firewall_check_access(paddr_t base, size_t size,
+				       const struct stm32_firewall_cfg *cfg)
+{
+	TEE_Result res = TEE_ERROR_ITEM_NOT_FOUND;
+	struct stm32_firewall_device *fdev = NULL;
+	uint32_t exceptions = 0;
+	unsigned int index = 0;
+
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+	fdev = find_device(base, size, &index);
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+
+	if (fdev && fdev->ops->has_access)
+		res = fdev->ops->has_access(fdev, index, base, size, cfg);
+
+	return res;
+}
+
+TEE_Result stm32_firewall_set_config(paddr_t base, size_t size,
+				     const struct stm32_firewall_cfg *cfg)
+{
+	TEE_Result res = TEE_ERROR_ITEM_NOT_FOUND;
+	struct stm32_firewall_device *fdev = NULL;
+	uint32_t exceptions = 0;
+	unsigned int index = 0;
+
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+	fdev = find_device(base, size, &index);
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+
+	if (fdev && fdev->ops->configure_access)
+		res = fdev->ops->configure_access(fdev, index, base, size, cfg);
+
+	return res;
+}
+
+struct stm32_firewall_device *stm32_firewall_dev_alloc(void)
+{
+	return calloc(1, sizeof(struct stm32_firewall_device));
+}
+
+void stm32_firewall_dev_free(struct stm32_firewall_device *fdev)
+{
+	uint32_t exceptions = 0;
+
+	assert(fdev);
+
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+	LIST_REMOVE(fdev, dev_list);
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+
+	free(fdev);
+}
+
+TEE_Result stm32_firewall_dev_register(struct stm32_firewall_device *fdev)
+{
+	uint32_t exceptions = 0;
+
+	assert(fdev && fdev->name && fdev->ops);
+
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+	LIST_INSERT_HEAD(&fw_dev_list, fdev, dev_list);
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result
+stm32_firewall_bus_dt_register(struct stm32_firewall_device *fdev,
+			       const void *fdt, int node)
+{
+	int idx = 0;
+	int len = 0;
+	int count = 0;
+	unsigned int id = 0;
+	const char *compat = NULL;
+	paddr_t pbase = 0;
+
+	/* Check secure access on main core */
+	static const struct stm32_firewall_cfg sec_cfg[] = {
+		{ FWLL_SEC_RW | FWLL_MASTER(0) },
+		{ }, /* Null terminated */
+	};
+
+	if (fdt_get_status(fdt, node) == DT_STATUS_DISABLED)
+		return TEE_SUCCESS;
+
+	count = fdt_stringlist_count(fdt, node, "compatible");
+	if (count < 0)
+		return TEE_SUCCESS;
+
+	pbase = fdt_reg_base_address(fdt, node);
+	if (pbase == DT_INFO_INVALID_REG)
+		return TEE_ERROR_GENERIC;
+
+	for (id = 0; id < fdev->compat->compat_size; id++)
+		if (fdev->compat->reg[id].addr == pbase)
+			break;
+
+	if (id == fdev->compat->compat_size)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (fdev->ops->has_access(fdev, id, pbase, 0, sec_cfg)) {
+		FMSG("Skip the %s is not secured",
+		     fdt_get_name(fdt, node, NULL));
+			return TEE_SUCCESS;
+	}
+
+	for (idx = 0; idx < count; idx++) {
+		compat = fdt_stringlist_get(fdt, node, "compatible", idx, &len);
+		assert(compat && len > 0);
+
+		add_probe_node_by_compat(fdt, node, compat);
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result stm32_firewall_bus_probe(struct stm32_firewall_device *fdev,
+				    const void *fdt, int node)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	int subnode = 0;
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		res = stm32_firewall_bus_dt_register(fdev, fdt, subnode);
+		if (res) {
+			DMSG("Failed on node %s with %#"PRIx32,
+			     fdt_get_name(fdt, subnode, NULL), res);
+			panic();
+		}
+	}
+
+	return TEE_SUCCESS;
+}
diff --git a/core/drivers/firewall/sub.mk b/core/drivers/firewall/sub.mk
new file mode 100644
index 000000000..29c5a97fd
--- /dev/null
+++ b/core/drivers/firewall/sub.mk
@@ -0,0 +1 @@
+srcs-$(CFG_STM32_FIREWALL) += stm32_firewall.c
diff --git a/core/drivers/sub.mk b/core/drivers/sub.mk
index 9e7625ae5..d3d7f8d5c 100644
--- a/core/drivers/sub.mk
+++ b/core/drivers/sub.mk
@@ -78,6 +78,7 @@ srcs-$(CFG_VERSAL_HUK) += versal_huk.c
 srcs-$(CFG_CBMEM_CONSOLE) += cbmem_console.c
 srcs-$(CFG_RISCV_PLIC) += plic.c
 subdirs-y += crypto
+subdirs-y += firewall
 subdirs-$(CFG_BNXT_FW) += bnxt
 subdirs-$(CFG_DRIVERS_CLK) += clk
 subdirs-$(CFG_DRIVERS_GPIO) += gpio
diff --git a/core/include/drivers/stm32_firewall.h b/core/include/drivers/stm32_firewall.h
new file mode 100644
index 000000000..db1351e00
--- /dev/null
+++ b/core/include/drivers/stm32_firewall.h
@@ -0,0 +1,212 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, STMicroelectronics - All Rights Reserved
+ */
+
+#ifndef __DRIVERS_STM32_FIREWALL_H
+#define __DRIVERS_STM32_FIREWALL_H
+
+#include <kernel/dt.h>
+#include <kernel/pm.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <sys/queue.h>
+#include <tee_api_types.h>
+#include <types_ext.h>
+
+/*
+ * Firewall configuration
+ *
+ * 32b configuration ID contains, from LSbits to MSbits:
+ * - 8 bit access permissions bit flags (rw, u/p, s/ns)
+ * - 8 bit attribute flags (locking)
+ * - 16 bit platform specific ID of the access master
+ */
+
+/* Master configuration bits */
+#define FWLL_MASTER_SHIFT	16
+#define FWLL_MASTER(x)		((x) << FWLL_MASTER_SHIFT)
+#define FWLL_MASTER_MASK	GENMASK_32(31, 16)
+
+/* Lock configuration */
+#define FWLL_CONF_LOCK		BIT(15)
+
+/* Access configuration */
+#define FWLL_CONF_MASK		GENMASK_32(7, 0)
+#define FWLL_NO_ACCESS		0
+#define FWLL_SEC_READ		BIT(0)
+#define FWLL_SEC_WRITE		BIT(1)
+#define FWLL_SEC_PRIV		BIT(2)
+#define FWLL_SEC_UNPRIV		BIT(3)
+#define FWLL_NSEC_READ		BIT(4)
+#define FWLL_NSEC_WRITE		BIT(5)
+#define FWLL_NSEC_PRIV		BIT(6)
+#define FWLL_NSEC_UNPRIV	BIT(7)
+
+#define FWLL_SEC_RW		(FWLL_SEC_READ | \
+				 FWLL_SEC_WRITE | \
+				 FWLL_SEC_PRIV | \
+				 FWLL_SEC_UNPRIV)
+
+#define FWLL_NSEC_RW		(FWLL_NSEC_READ | \
+				 FWLL_NSEC_WRITE | \
+				 FWLL_NSEC_PRIV | \
+				 FWLL_NSEC_UNPRIV)
+
+#define FWLL_FULL_ACCESS	(FWLL_SEC_RW | \
+				 FWLL_NSEC_RW)
+
+struct stm32_firewall_device;
+
+/**
+ * @brief struct stm32_firewall_reg - define the firewall device management
+ *
+ * @addr : controlled physical address.
+ * @size : length of the controlled area.
+ */
+struct stm32_firewall_reg {
+	paddr_t addr;
+	size_t size;
+};
+
+/**
+ * @brief struct stm32_firewall_compat - Device tree compatibility table
+ *
+ * @reg: stm32_firewall_reg structure.
+ * @compat_size: Array size of the stm32_firewall_reg structure.
+ */
+struct stm32_firewall_compat {
+	struct stm32_firewall_reg *reg;
+	size_t compat_size;
+};
+
+/**
+ * @brief struct stm32_firewall_cfg - Firewall configuration structure
+ *
+ * @access: Bit flags for permission, FWLL_[N]SEC_*
+ */
+struct stm32_firewall_cfg {
+	uint32_t access;
+};
+
+/**
+ * @brief struct stm32_firewall_ops
+ *
+ * @has_access: Check the access regarding the requested configuration.
+ * @configure_access: Configure access based on requested configuration.
+ * @request_access: Lock access for the requester.
+ * @release_access: Unlock access of the requester.
+ */
+struct stm32_firewall_ops {
+	TEE_Result (*has_access)(struct stm32_firewall_device *fdev,
+				 unsigned int id,
+				 paddr_t addr, size_t sz,
+				 const struct stm32_firewall_cfg *cfg);
+	TEE_Result (*configure_access)(struct stm32_firewall_device *fdev,
+				       unsigned int id,
+				       paddr_t addr, size_t sz,
+				       const struct stm32_firewall_cfg *cfg);
+	TEE_Result (*request_access)(struct stm32_firewall_device *fdev,
+				     unsigned int id,
+				     paddr_t addr, size_t sz,
+				     const struct stm32_firewall_cfg *cfg);
+	TEE_Result (*release_access)(struct stm32_firewall_device *fdev,
+				     unsigned int id,
+				     paddr_t addr, size_t sz,
+				     const struct stm32_firewall_cfg *cfg);
+};
+
+/**
+ * @brief struct stm32_firewall_device
+ *
+ * @name: device name.
+ * @ops: Operation table of the firewall.
+ * @compat: List of the devices controlled by the firewall instance
+ * @priv: Optional private data supplied by driver.
+ * @dev_list: Firewall instance reference registered firewall list
+ */
+struct stm32_firewall_device {
+	const char *name;
+	const struct stm32_firewall_ops *ops;
+	const struct stm32_firewall_compat *compat;
+	void *priv;
+
+	LIST_ENTRY(stm32_firewall_device) dev_list;
+};
+
+/* API for provider */
+
+/**
+ * @brief stm32_firewall_priv - Retrieve firewall private data
+ *
+ * @fdev: stm32_firewall_device struct to be freed.
+ *
+ * Return firewall private data
+ */
+static inline void *stm32_firewall_priv(struct stm32_firewall_device *fdev)
+{
+	return fdev->priv;
+}
+
+/**
+ * @brief stm32_firewall_check_access - Check the configuration access
+ * regarding a defined range.
+ *
+ * @base: Physical address
+ * @size: Size to be checked
+ * @cfg: Tested configuration
+ *
+ * Returns TEE_SUCCESS if access is permitted, or an error code otherwise
+ */
+TEE_Result stm32_firewall_check_access(paddr_t base, size_t size,
+				       const struct stm32_firewall_cfg *cfg);
+
+/**
+ * @brief stm32_firewall_set_config - Define the configuration access
+ * regarding a defined range.
+ *
+ * @base: Physical address
+ * @size: Size to be checked
+ * @cfg: Tested configuration
+ *
+ * Returns a TEE_Result compliant value.
+ */
+TEE_Result stm32_firewall_set_config(paddr_t base, size_t size,
+				     const struct stm32_firewall_cfg *cfg);
+
+/**
+ * @brief stm32_firewall_dev_alloc - Allocate a new firewall device.
+ *
+ * Returns a stm32_firewall_device struct or NULL if allocation failed.
+ */
+struct stm32_firewall_device *stm32_firewall_dev_alloc(void);
+
+/**
+ * @brief stm32_firewall_dev_free - Free a firewall device.
+ *
+ * @fdev: stm32_firewall_device struct to be freed.
+ */
+void stm32_firewall_dev_free(struct stm32_firewall_device *fdev);
+
+/**
+ * @brief stm32_firewall_dev_register - Register a firewall device
+ * list.
+ *
+ * @fdev: stm32_firewall_device struct to be added.
+ *
+ * Returns a TEE_Result compliant value.
+ */
+TEE_Result stm32_firewall_dev_register(struct stm32_firewall_device *fdev);
+
+/**
+ * @brief stm32_firewall_bus_probe - Probe a firewall bus driver child.
+ *
+ * @fdev: firewall_device struct to be probed.
+ * @fdt: device tree address.
+ * @node: firewall bus node.
+ *
+ * Return a TEE_Result compliant value.
+ */
+TEE_Result stm32_firewall_bus_probe(struct stm32_firewall_device *fdev,
+				    const void *fdt, int node);
+#endif /* __DRIVERS_STM32_FIREWALL_H */
-- 
2.47.2

